keyboard configuration crashes on Android

Reporting a crash for me (Lineage OS 18.1 + pixelwheels latest nightly build 20211216) when pressing the keyboard configure button:
```
12-16 16:34:37.974 24120 25343 E AndroidRuntime: FATAL EXCEPTION: GLThread 1395
12-16 16:34:37.974 24120 25343 E AndroidRuntime: Process: com.agateau.tinywheels.android, PID: 24120
12-16 16:34:37.974 24120 25343 E AndroidRuntime: java.lang.RuntimeException: Actor: ButtonMenuItem: CONFIGURE
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.badlogic.gdx.scenes.scene2d.Actor.notify(Actor.java:192)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.badlogic.gdx.scenes.scene2d.Actor.fire(Actor.java:152)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.agateau.ui.menu.MenuItemTriggerEvent.fire(MenuItemTriggerEvent.java:30)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.agateau.ui.menu.ButtonMenuItem$1.clicked(ButtonMenuItem.java:53)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.badlogic.gdx.scenes.scene2d.utils.ClickListener.touchUp(ClickListener.java:89)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.badlogic.gdx.scenes.scene2d.InputListener.handle(InputListener.java:69)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.badlogic.gdx.scenes.scene2d.Stage.touchUp(Stage.java:352)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.badlogic.gdx.InputMultiplexer.touchUp(InputMultiplexer.java:124)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.badlogic.gdx.backends.android.DefaultAndroidInput.processEvents(DefaultAndroidInput.java:413)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.badlogic.gdx.backends.android.AndroidGraphics.onDrawFrame(AndroidGraphics.java:472)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at android.opengl.GLSurfaceView$GLThread.guardedRun(GLSurfaceView.java:1573)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at android.opengl.GLSurfaceView$GLThread.run(GLSurfaceView.java:1272)
12-16 16:34:37.974 24120 25343 E AndroidRuntime: Caused by: java.lang.ClassCastException: com.agateau.pixelwheels.gameinput.PieTouchInputHandler cannot be cast to com.agateau.pixelwheels.gameinput.KeyboardInputHandler
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.agateau.pixelwheels.screens.KeyboardConfigScreen.<init>(KeyboardConfigScreen.java:72)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.agateau.pixelwheels.screens.-$$Lambda$oZI4LdR43zsXaK0cTvG3IPOhtSc.createScreen(Unknown Source:2)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.agateau.pixelwheels.screens.ConfigScreen$InputSelectorController.onConfigureClicked(ConfigScreen.java:354)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.agateau.pixelwheels.screens.ConfigScreen$InputSelectorController.access$400(ConfigScreen.java:302)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.agateau.pixelwheels.screens.ConfigScreen$InputSelectorController$2.triggered(ConfigScreen.java:337)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.agateau.ui.menu.MenuItemListener.handle(MenuItemListener.java:33)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        at com.badlogic.gdx.scenes.scene2d.Actor.notify(Actor.java:188)
12-16 16:34:37.974 24120 25343 E AndroidRuntime:        ... 11 more

```

_Originally posted in https://github.com/agateau/pixelwheels/issues/172#issuecomment-995557021_

--- a/core/src/com/agateau/pixelwheels/gameinput/GamepadInputHandler.java
+++ b/core/src/com/agateau/pixelwheels/gameinput/GamepadInputHandler.java
@@ -25,7 +25,7 @@ import com.agateau.ui.GamepadInputMappers;
 import com.badlogic.gdx.utils.Array;
 
 /** Handle gamepad input, for desktop mode */
-public class GamepadInputHandler extends KeyboardInputHandler {
+public class GamepadInputHandler extends InputMapperInputHandler {
 
     public static class Factory implements GameInputHandlerFactory {
         final Array<GameInputHandler> mHandlers = new Array<>();

--- a/core/src/com/agateau/pixelwheels/gameinput/InputMapperInputHandler.java
+++ b/core/src/com/agateau/pixelwheels/gameinput/InputMapperInputHandler.java
@@ -1,0 +1,80 @@
+/*
+ * Copyright 2021 Aurélien Gâteau <mail@agateau.com>
+ *
+ * This file is part of Pixel Wheels.
+ *
+ * Pixel Wheels is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free
+ * Software Foundation, either version 3 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.agateau.pixelwheels.gameinput;
+
+import com.agateau.pixelwheels.Assets;
+import com.agateau.pixelwheels.GamePlay;
+import com.agateau.pixelwheels.bonus.Bonus;
+import com.agateau.pixelwheels.racescreen.Hud;
+import com.agateau.ui.InputMapper;
+import com.agateau.ui.VirtualKey;
+import com.badlogic.gdx.Preferences;
+import com.badlogic.gdx.math.MathUtils;
+
+/** Base class for InputMapper-based GameInputHandlers */
+public class InputMapperInputHandler implements GameInputHandler {
+    private final InputMapper mInputMapper;
+    private final GameInput mInput = new GameInput();
+
+    InputMapperInputHandler(InputMapper inputMapper) {
+        mInputMapper = inputMapper;
+    }
+
+    @Override
+    public GameInput getGameInput() {
+        mInput.braking = mInputMapper.isKeyPressed(VirtualKey.DOWN);
+        mInput.accelerating = !mInput.braking;
+        if (mInputMapper.isKeyPressed(VirtualKey.LEFT)) {
+            mInput.direction += GamePlay.instance.steeringStep;
+        } else if (mInputMapper.isKeyPressed(VirtualKey.RIGHT)) {
+            mInput.direction -= GamePlay.instance.steeringStep;
+        } else {
+            mInput.direction = 0;
+        }
+        mInput.direction = MathUtils.clamp(mInput.direction, -1, 1);
+        mInput.triggeringBonus = mInputMapper.isKeyPressed(VirtualKey.TRIGGER);
+
+        return mInput;
+    }
+
+    @Override
+    public void loadConfig(Preferences preferences, String prefix, int playerIdx) {
+        mInputMapper.loadConfig(preferences, prefix, playerIdx);
+    }
+
+    @Override
+    public void saveConfig(Preferences preferences, String prefix) {
+        mInputMapper.saveConfig(preferences, prefix);
+    }
+
+    @Override
+    public void createHudButtons(Assets assets, Hud hud) {}
+
+    @Override
+    public void setBonus(Bonus bonus) {}
+
+    @Override
+    public boolean isAvailable() {
+        return mInputMapper.isAvailable();
+    }
+
+    public InputMapper getInputMapper() {
+        return mInputMapper;
+    }
+}

--- a/core/src/com/agateau/pixelwheels/gameinput/KeyboardInputHandler.java
+++ b/core/src/com/agateau/pixelwheels/gameinput/KeyboardInputHandler.java
@@ -20,20 +20,12 @@ package com.agateau.pixelwheels.gameinput;
 
 import static com.agateau.translations.Translator.tr;
 
-import com.agateau.pixelwheels.Assets;
 import com.agateau.pixelwheels.Constants;
-import com.agateau.pixelwheels.GamePlay;
-import com.agateau.pixelwheels.bonus.Bonus;
-import com.agateau.pixelwheels.racescreen.Hud;
-import com.agateau.ui.InputMapper;
 import com.agateau.ui.KeyMapper;
-import com.agateau.ui.VirtualKey;
-import com.badlogic.gdx.Preferences;
-import com.badlogic.gdx.math.MathUtils;
 import com.badlogic.gdx.utils.Array;
 
 /** Handle keyboard input, for desktop mode */
-public class KeyboardInputHandler implements GameInputHandler {
+public class KeyboardInputHandler extends InputMapperInputHandler {
     public static class Factory implements GameInputHandlerFactory {
         final Array<GameInputHandler> mHandlers = new Array<>();
 
@@ -60,52 +52,11 @@ public class KeyboardInputHandler implements GameInputHandler {
         }
     }
 
-    private final InputMapper mInputMapper;
-    private final GameInput mInput = new GameInput();
-
-    KeyboardInputHandler(InputMapper inputMapper) {
-        mInputMapper = inputMapper;
+    KeyboardInputHandler(KeyMapper keyMapper) {
+        super(keyMapper);
     }
 
-    @Override
-    public GameInput getGameInput() {
-        mInput.braking = mInputMapper.isKeyPressed(VirtualKey.DOWN);
-        mInput.accelerating = !mInput.braking;
-        if (mInputMapper.isKeyPressed(VirtualKey.LEFT)) {
-            mInput.direction += GamePlay.instance.steeringStep;
-        } else if (mInputMapper.isKeyPressed(VirtualKey.RIGHT)) {
-            mInput.direction -= GamePlay.instance.steeringStep;
-        } else {
-            mInput.direction = 0;
-        }
-        mInput.direction = MathUtils.clamp(mInput.direction, -1, 1);
-        mInput.triggeringBonus = mInputMapper.isKeyPressed(VirtualKey.TRIGGER);
-
-        return mInput;
-    }
-
-    @Override
-    public void loadConfig(Preferences preferences, String prefix, int playerIdx) {
-        mInputMapper.loadConfig(preferences, prefix, playerIdx);
-    }
-
-    @Override
-    public void saveConfig(Preferences preferences, String prefix) {
-        mInputMapper.saveConfig(preferences, prefix);
-    }
-
-    @Override
-    public void createHudButtons(Assets assets, Hud hud) {}
-
-    @Override
-    public void setBonus(Bonus bonus) {}
-
-    @Override
-    public boolean isAvailable() {
-        return mInputMapper.isAvailable();
-    }
-
-    public InputMapper getInputMapper() {
-        return mInputMapper;
+    public KeyMapper getKeyMapper() {
+        return (KeyMapper) getInputMapper();
     }
 }

--- a/core/src/com/agateau/pixelwheels/screens/KeyboardConfigScreen.java
+++ b/core/src/com/agateau/pixelwheels/screens/KeyboardConfigScreen.java
@@ -24,6 +24,7 @@ import com.agateau.pixelwheels.Constants;
 import com.agateau.pixelwheels.GameConfig;
 import com.agateau.pixelwheels.PwGame;
 import com.agateau.pixelwheels.PwRefreshHelper;
+import com.agateau.pixelwheels.gameinput.GameInputHandler;
 import com.agateau.pixelwheels.gameinput.KeyboardInputHandler;
 import com.agateau.ui.KeyMapper;
 import com.agateau.ui.VirtualKey;
@@ -68,12 +69,12 @@ public class KeyboardConfigScreen extends PwStageScreen {
         // Find the current player KeyMapper, but also get all KeyMappers: we need them
         // to check for conflicts
         for (int idx = 0; idx < Constants.MAX_PLAYERS; ++idx) {
-            KeyboardInputHandler handler =
-                    (KeyboardInputHandler) mGame.getConfig().getPlayerInputHandler(idx);
-            if (handler == null) {
+            GameInputHandler handler = mGame.getConfig().getPlayerInputHandler(idx);
+            if (!(handler instanceof KeyboardInputHandler)) {
                 continue;
             }
-            KeyMapper mapper = (KeyMapper) handler.getInputMapper();
+            KeyboardInputHandler keyboardInputHandler = (KeyboardInputHandler) handler;
+            KeyMapper mapper = keyboardInputHandler.getKeyMapper();
             mKeyMappers.add(mapper);
             if (idx == playerIdx) {
                 mKeyMapper = mapper;

